Refine the set of NullableAnnotation values to the three speakable options

gafter opened the pull request:
- Eliminate values of `NullableAnnotation` that are "unspeakable"
- Change the result of dynamic invocations to result in an oblivious state
- Simplify the code in a few places where we compute result annotations (TSWA and `TypeWithState` are capable of storing an annotation/state even for typeless expressions)

Fixes #33638 
Filed #33924 to fix the result state of `?:`

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs:
```diff
@@ -49,194 +47,96 @@ internal enum NullableFlowState : byte
-        /// This method projects nullable annotations onto a smaller set that can be expressed in source.
-        /// </summary>
-        public static NullableAnnotation AsSpeakable(this NullableAnnotation annotation, TypeSymbol type)
+        public static bool MaybeNull(this NullableFlowState state)
```
>public static bool MaybeNull(this NullableFlowState state) [](start = 8, length = 58)

nit: this extension should be in `NullableFlowStateExtensions` not `NullableAnnotationExtensions`.

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs:
```diff
@@ -49,194 +47,96 @@ internal enum NullableFlowState : byte
-
-            return NullableAnnotation.Annotated;
-        }
+        private static int[] _permute = new[] { 1, 0, 2 };
```
>private static int[] _permute = new[] { 1, 0, 2 }; [](start = 8, length = 50)

This trick feels too clever (ie. hard to review). I think we should either write the Meet/Join in an expanded form, or we should re-order the enum entries.

gafter replied:
I would love to reorder, but we have plenty of hand-written IL with hardcoded attribute values with the enum values hardcoded.  That would be a fair amount of work to fix up the tests but I think worth it.  Opened [https://github.com/dotnet/roslyn/issues/33952](https://github.com/dotnet/roslyn/issues/33952) to track this.

---
In reply to: [263508874](https://github.com/dotnet/roslyn/pull/33927#discussion_r263508874) [](ancestors = 263508874)

jcouv replied:
For what it's worth, my suggestion is to re-order the enum, but keep the IL that we emit as-is (otherwise we'll be causing unnecessary surprises to early adopters)

---
In reply to: [263598753](https://github.com/dotnet/roslyn/pull/33927#discussion_r263598753) [](ancestors = 263598753,263508874)

gafter replied:
@cston points out that reordering would be a breaking change.  It is really sad to be constrained by breaks at this juncture.

---
In reply to: [263598753](https://github.com/dotnet/roslyn/pull/33927#discussion_r263598753) [](ancestors = 263598753,263508874)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs:
```diff
@@ -324,25 +224,6 @@ private TypeSymbolWithAnnotations(TypeSymbol defaultType, NullableAnnotation nul
             _extensions = extensions;
         }
 
-        public TypeSymbolWithAnnotations AsSpeakable()
```
>AsSpeakable [](start = 41, length = 11)

note: there is a mention of "unspeakable" in the design doc. Could you remove it too? Thanks

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs:
```diff
@@ -577,14 +446,14 @@ public string ToDisplayString(SymbolDisplayFormat format = null)
-                    (NullableAnnotation == NullableAnnotation.Annotated ||
-                     (NullableAnnotation == NullableAnnotation.Nullable && !TypeSymbol.IsTypeParameterDisallowingAnnotation())))
+                    (NullableAnnotation.IsAnnotated() ||
+                     (NullableAnnotation.IsAnnotated() && !TypeSymbol.IsTypeParameterDisallowingAnnotation())))
```
>NullableAnnotation.IsAnnotated() [](start = 22, length = 32)

redundant check

jcouv commented:
>             NullableAnnotation mergeNullableAnnotation(TypeSymbol type, NullableAnnotation a, NullableAnnotation b)

I think `TSWA.MergeNullableAnnotation` can be used here, and this local function could be deleted.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:1520 in 0178fec. [](commit_id = 0178fec154162c000296962a1fe65a24b8cfc756, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:
```diff
@@ -98,13 +98,13 @@ private bool ShouldAddNullableAnnotation(TypeSymbolWithAnnotations typeOpt)
                 !typeOpt.IsNullableType() && !typeOpt.IsValueType &&
                 (typeOpt.NullableAnnotation == NullableAnnotation.Annotated ||
-                 (typeOpt.NullableAnnotation == NullableAnnotation.Nullable && !typeOpt.TypeSymbol.IsTypeParameterDisallowingAnnotation())))
+                 (typeOpt.NullableAnnotation == NullableAnnotation.Annotated && !typeOpt.TypeSymbol.IsTypeParameterDisallowingAnnotation())))
```
>typeOpt.NullableAnnotation == NullableAnnotation.Annotated [](start = 18, length = 58)

nit: consider using `IsAnnotated()` to check

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -2293,7 +2262,7 @@ public override BoundNode VisitConditionalOperator(BoundConditionalOperator node
                 {
                     visitResult = TypeSymbolWithAnnotations.Create(resultType,
-                                                                   convertedConsequenceResult.NullableAnnotation.JoinForFixingLowerBounds(convertedAlternativeResult.NullableAnnotation));
+                                                                   convertedConsequenceResult.NullableAnnotation.Join(convertedAlternativeResult.NullableAnnotation));
```
>Join [](start = 113, length = 4)

not related to this PR: consider fixing atypical formatting

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -73792,14 +73783,13 @@ static void F(C<object> x, C<object?>? y)
         private readonly static NullableFlowState[] s_AllNullableFlowStates = (NullableFlowState[])Enum.GetValues(typeof(NullableFlowState));
 
         [Fact]
         public void TestJoinForFixingLowerBounds()
```
>TestJoinForFixingLowerBounds [](start = 20, length = 28)

this test name should be adjusted. Same for new few tests below

gafter commented:
@jcouv I think I have responded to all of your comments.
@dotnet/roslyn-compiler May I please have another review?

cston commented on src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs:
```diff
@@ -0,0 +1,37 @@
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
```
Consider removing unnecessary entries.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotation.cs:
```diff
@@ -0,0 +1,25 @@
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// The nullable annotations that can apply in source.
```
Consider adding info in the comment on how this works with types nested in a generic.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotation.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <summary>
+    /// The nullable annotations that can apply in source.
+    /// </summary>
+    internal enum NullableAnnotation : byte
```
Consider adding a comment that order is important here as the code base will do logic comparisons on the numeric value.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotation.cs:
```diff
@@ -0,0 +1,25 @@
+        /// <summary>
+        /// No information. Think oblivious.
+        /// </summary>
+        Unknown,
```
Why not `Oblivious` here? All the helpers that use the value use this term.

jcouv replied:
For a while we thought we should use "unknown" instead of "oblivious". 
But now, Mads' spec bolsters the term "oblivious", so we may need to revisit.

We have a tracking issue already https://github.com/dotnet/roslyn/issues/29632 (this affects docs and public API)

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/NullableFlowStateExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    internal static class NullableFlowStateExtensions
+    {
+        public static bool MaybeNull(this NullableFlowState state) => state == NullableFlowState.MaybeNull;
```
`IsMaybeNull`

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs:
```diff
@@ -0,0 +1,37 @@
+        public TypeSymbol Type { get; }
+        public NullableFlowState State { get; }
+        public bool HasNullType => Type is null;
+        public bool MaybeNull => State == NullableFlowState.MaybeNull;
```
`IsMaybeNull`

gafter replied:
`MayBeNull`?

---
In reply to: [263853738](https://github.com/dotnet/roslyn/pull/33927#discussion_r263853738) [](ancestors = 263853738)

jaredpar replied:
Discussed in person `IsMaybeNull` doesn't make a lot of sense when you say it out loud. Keeping as is.

jaredpar commented:
Done with review pass (iteration 7). Really like where this change is taking us.

gafter commented:
@jaredpar @jcouv @cston I believe I have addressed all comments.  Also, I have renamed "Unknown" to "Oblivious" so please review the latest iteration to see if you approve.
